Common/include/toolboxes/ndflattener.hpp

```diff
@@ -0,0 +1,846 @@
+      if(++i<to) output << ", ";
+    }
+    output << "]";
+  }
+
+public:
+
+  /*! \brief Basic constructor. Afterwards, initialization can be done with initialize_or_refresh.
+   * 
+   * Called recursively when a derived class (higher K) is constructed.
+   */
+  NdFlattener() {};
+
+  /*! \brief Constructor which calls initialize_or_refresh.
+   *
+   */
+  template<class... ARGS>
+  NdFlattener(ARGS... args) {
+    initialize_or_refresh(args...);
+  }
+
+  /*! \brief Initialize or refresh the NdFlattener.
+   * \details Either a 'recursive function' or 'collective communication'
+   * may be used. When the NdFlattener does not hold data yet, it is 
+   * initialized, meaning that the data are collected and the indices arrays
+   * are allocated and filled. Otherwise it is refreshed, meaning that the data are
+   * recollected under the assumption that the indices arrays did not change.
+   */
+  template<class ...ARGS>
+  void initialize_or_refresh(ARGS... args){
+    if( initialized() ){
+      refresh(args...);
+    } else {
+      initialize(args...);
+    }
+  }
+
+  /*! \brief Initialization status of the NdFlattener.
+   * \returns true if the NdFlattener has been initialized
+   */
+  bool initialized(){
+    return nNodes>0;
+  }
+
+protected:
+  /*! \brief Allocate the indices array after \a nNodes has been determined.
+   */
+  void allocate() {
+    indices.reserve(nNodes+1);
+    indices[0] = 0;
+    Base::allocate();
+  }
+
+  /*! \brief Set \a iNode to 0 in all layers.
+   */
+  void reset_iNode(){
+    iNode = 0;
+    Base::reset_iNode();
+  }
+
+  /*=== Construct from 'recursive function' ===*/
+public:
+  /*! \brief Initialize from a 'recursive function'.
+   *
+   * The function should return a pair. Its first entry is the number of children. 
+   * Its second entry is a function with the same meaning, recursively 
+   * one layer down.
+   * \param f - the 'recursive function'
+   */
+  template<class f_type>
+  void initialize(f_type f) {
+    count_f(f);
+    allocate();
+    set_f(f, false);
+  }
+
+  /*! \brief Refresh the data according to the 'recursive function'
+   *
+   * The NdFlattener must have been constructed with a 'recursive function'.
+   * Now refresh the values with another 'recursive function'. The subarray lengths 
+   * resulting from both 'recursive functions' must coincide, as the indices arrays
+   * are not changed.
+   * 
+   * \param f - the 'recursive function'
+   * \tparam f_type - to allow for any type of the 'recursive function'
+   */
+  template<class f_type>
+  void refresh(f_type f){
+    reset_iNode();
+    set_f(f, true);
+  }
+
+protected:
+  /*! \brief Determine the space required for reading the 'recursive function'.
+   *
+   * \param f - the 'recursive function'
+   */
+  template<class f_type>
+  void count_f(f_type f) {
+    Index nChild = f.first;
+    for(Index iChild=0; iChild<nChild; iChild++){
+      nNodes++;
```

Comment:
nNodes = nChild ?

Reply:
No. `nNodes` is the length of the indices (or data) array (+-1), `nChild` is the number of indices (or data elements) "below" a particular partial index tuple. The former is the sum of the latter, so to speak.

Reply:
Ok it is only equal for the top level.